server/src/main/java/org/elasticsearch/cluster/routing/allocation/WriteLoadConstraintMonitor.java

```diff
@@ -99,6 +132,26 @@ public void onNewInfo(ClusterInfo clusterInfo) {
             }
         }
 
+        // reset hotspotNodeStartTimes if the term has changed
+        if (state.term() != hotspotNodeStartTimesLastTerm) {
+            hotspotNodeStartTimesLastTerm = state.term();
+            hotspotNodeStartTimes.clear();
+        }
+
+        final long currentTimeMillis = currentTimeMillisSupplier.getAsLong();
+        Set<String> lastHotspotNodes = hotspotNodeStartTimes.keySet();
+        Set<String> staleHotspotNodes = Sets.difference(lastHotspotNodes, writeNodeIdsExceedingQueueLatencyThreshold);
+
+        for (String nodeId : staleHotspotNodes) {
+            assert hotspotNodeStartTimes.containsKey(nodeId) : "Map should contain key from its own subset";
+            long hotspotStartTime = hotspotNodeStartTimes.remove(nodeId);
+            long hotspotDuration = currentTimeMillis - hotspotStartTime;
+            assert hotspotDuration >= 0 : "hotspot duration should always be non-negative";
+            hotspotDurationHistogram.record(hotspotDuration / 1000.0);
+        }
+        hotspotNodesCount = hotspotNodeStartTimes.size();
+        hotspotNodesCountUpdatedSinceLastRead = true;
```

Comment:
I wonder if it'd be feasible to break this stuff out into a method, just so it doesn't distract too much from the monitoring logic?

Likewise with the recording of the start times?